File: Labs/Lab-04-20231013/app/src/main/java/sap/escooters/domain_layer/Ride.java
package sap.escooters.domain_layer;

import java.util.Date;
import java.util.Optional;

import io.vertx.core.json.JsonObject;

public class Ride {

	private Date startedDate;
	private Optional<Date> endDate;
	private User user;
	private EScooter scooter;
	private boolean ongoing;
	private String id;
	
	public Ride(String id, User user, EScooter scooter) {
		this.id = id;
		this.startedDate = new Date();
		this.endDate = Optional.empty();
		this.user = user;
		this.scooter = scooter;
		ongoing = true;
	}
	
	public String getId() {
		return id;
	}
	
	public void end() {
		endDate = Optional.of(new Date());
		ongoing = false;
		save();
	}

	public Date getStartedDate() {
		return startedDate;
	}

	public boolean isOngoing() {
		return this.ongoing;
	}
	
	public Optional<Date> getEndDate() {
		return endDate;
	}

	public User getUser() {
		return user;
	}

	public EScooter getEScooter() {
		return scooter;
	}
	
	public void save() {
		try {
			DomainLayerImpl.getDataSourceLayer().saveRide(toJson());
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}
	
	public JsonObject toJson() {
		JsonObject rideObj = new JsonObject();
		rideObj.put("id", this.getId());
		rideObj.put("userId", this.getUser().getId());
		rideObj.put("escooterId", this.getEScooter().getId());
		rideObj.put("startDate", this.getStartedDate().toString());
		Optional<Date> endDate = this.getEndDate();
		
		if (endDate.isPresent()) {
			rideObj.put("endDate", endDate.get().toString());			
		} else {
			rideObj.putNull("location");			
		}			
		return rideObj;
	}

	
}

File: Labs/Lab-05-20231016/app/src/main/kotlin/assigment03/escooters/domain_layer/Ride.kt
package assigment03.escooters.domain_layer

import io.vertx.core.json.JsonObject
import java.util.*

class Ride(val id: String, val user: User, val scooter: EScooter) {
    private val startedDate = Date()
    private var endDate: Date? = null
    var ongoing = true
        private set

    fun end() {
        endDate = Date()
        ongoing = false
        save()
    }

    fun save() {
        try {
            DomainLayerImpl.getDataSourceLayer()?.saveRide(toJson())
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun toJson() = JsonObject().apply {
        put("id", id)
        put("userId", user.id)
        put("escooterId", scooter.id)
        put("startDate", startedDate.toString())
        endDate?.let {
            put("endDate", it.toString())
        } ?: putNull("location")
    }
}





